### PR TITLE
Refactor the DHCP client

### PR DESCRIPTION
**Pull request by yaocw2020**
- Refactor the DHCP client using `github.com/insomniacslk/dhcp/dhcpv4`
- When requesting IP through DHCP,  we store the hardware address of the macvlan interface and IPv4 address in service annotation in order to reuse them.

issue: https://github.com/kube-vip/kube-vip/issues/220

Signed-off-by: yaocw2020 <yaocanwu@gmail.com>

### CODE REVIEW

File: pkg/vip/ddns.go [hunk 1]
```diff
@@ -41,14 +42,9 @@ func (ddns *ddnsManager) Start() (string, error) {
...
-		// send host name, not current os.Hostname, but configured name from user
-		Hostname:  ddns.network.DDNSHostName(),
-		Interface: iface,
-		OnBound: func(lease *Lease) {
-			ipCh <- lease.ClientIP
-		},
-	}
+	client := NewDHCPClient(iface, false, "", func(lease *nclient4.Lease) {
+		ipCh <- lease.ACK.YourIPAddr.String()
+	})
```
Comment (yaocw2020) [inline pkg/vip/ddns.go:47]: I do not familiar with ddns and not sure how to use a DHCP client in ddns. @thebsdbox

**Comment by thebsdbox** [comment]
wow.. That's a big one! 

I've only just re-implemented DHCP too 😢

File: pkg/vip/ddns.go (hunk 1 above)
Comment (thebsdbox) [reply pkg/vip/ddns.go:47]: This looks OK to me.

**Comment by yaocw2020** [comment]
> wow.. That's a big one!
> 
> I've only just re-implemented DHCP too 😢

I know that. I have tried my best to meet the requirements base on your implementation but I failed. It's forced to re-implement the DHCP client. :dizzy_face:

**Comment by thebsdbox** [comment]
Looks like there are some build errors in it's current state?

**Comment by yaocw2020** [comment]
> Looks like there are some build errors in it's current state?

I will fix them tomorrow. I find that there are some duplicated codes between `pkg/manager` and `pkg/service`. It confuses me.

**Comment by yaocw2020** [comment]
> Signed-off-by: yaocw2020 [yaocanwu@gmail.com](mailto:yaocanwu@gmail.com)

**Comment by thebsdbox** [comment]
Passes the tests.. going to take it for a test on a cluster and if alls well, we should be good to merge it.

**Comment by thebsdbox** [comment]
```
dan@k8s01:~$  kubectl apply -f https://raw.githubusercontent.com/kube-vip/kube-vip-cloud-provider/main/manifest/kube-vip-cloud-controller.yaml
serviceaccount/kube-vip-cloud-controller created
clusterrole.rbac.authorization.k8s.io/system:kube-vip-cloud-controller-role created
clusterrolebinding.rbac.authorization.k8s.io/system:kube-vip-cloud-controller-binding created
statefulset.apps/kube-vip-cloud-provider created
dan@k8s01:~$ kubectl apply -f https://k8s.io/examples/application/deployment.yaml
deployment.apps/nginx-deployment created
dan@k8s01:~$ kubectl expose deployment nginx-deployment --port=80 --type=LoadBalancer --name=nginx-dhcp --load-balancer-ip=0.0.0.0
service/nginx-dhcp exposed
dan@k8s01:~$ kubectl get svc
NAME         TYPE           CLUSTER-IP       EXTERNAL-IP     PORT(S)        AGE
kubernetes   ClusterIP      10.96.0.1        <none>          443/TCP        9m36s
nginx-dhcp   LoadBalancer   10.101.236.225   192.168.0.121   80:32204/TCP   4s
dan@k8s01:~$ alias k=kubectl
dan@k8s01:~$ k get pods -A
NAMESPACE     NAME                                       READY   STATUS    RESTARTS   AGE
default       nginx-deployment-6b474476c4-z7cbg          1/1     Running   0          62s
default       nginx-deployment-6b474476c4-zrtxk          1/1     Running   0          62s
kube-system   calico-kube-controllers-5978c5f6b5-2vks7   1/1     Running   0          10m
kube-system   calico-node-2qksk                          1/1     Running   0          9m2s
kube-system   calico-node-9v4kh                          1/1     Running   0          10m
kube-system   calico-node-grr4w                          1/1     Running   0          8m31s
kube-system   calico-node-pkg7s                          1/1     Running   0          10m
kube-system   calico-node-pkhrr                          1/1     Running   0          8m41s
kube-system   coredns-66bff467f8-h2bmq                   1/1     Running   0          10m
kube-system   coredns-66bff467f8-j778b                   1/1     Running   0          10m
kube-system   etcd-k8s01                                 1/1     Running   0          10m
kube-system   etcd-k8s02                                 1/1     Running   0          9m44s
kube-system   etcd-k8s03                                 1/1     Running   0          7m55s
kube-system   kube-apiserver-k8s01                       1/1     Running   0          10m
kube-system   kube-apiserver-k8s02                       1/1     Running   0          10m
kube-system   kube-apiserver-k8s03                       1/1     Running   1          8m3s
kube-system   kube-controller-manager-k8s01              1/1     Running   1          10m
kube-system   kube-controller-manager-k8s02              1/1     Running   0          10m
kube-system   kube-controller-manager-k8s03              1/1     Running   0          7m57s
kube-system   kube-proxy-67bfj                           1/1     Running   0          10m
kube-system   kube-proxy-7zmpc                           1/1     Running   0          8m41s
kube-system   kube-proxy-9bq7x                           1/1     Running   0          9m2s
kube-system   kube-proxy-kzhds                           1/1     Running   0          8m31s
kube-system   kube-proxy-vmxw7                           1/1     Running   0          10m
kube-system   kube-scheduler-k8s01                       1/1     Running   1          10m
kube-system   kube-scheduler-k8s02                       1/1     Running   0          10m
kube-system   kube-scheduler-k8s03                       1/1     Running   0          8m16s
kube-system   kube-vip-cloud-provider-0                  1/1     Running   0          80s
kube-system   kube-vip-k8s01                             1/1     Running   1          10m
kube-system   kube-vip-k8s02                             1/1     Running   0          9m40s
kube-system   kube-vip-k8s03                             1/1     Running   0          8m49s
dan@k8s01:~$ k logs -n kube-system   kube-vip-k8s01
time="2021-07-07T08:29:25Z" level=info msg="server started"
time="2021-07-07T08:29:25Z" level=info msg="Starting Kube-vip Manager with the ARP engine"
time="2021-07-07T08:29:25Z" level=info msg="Namespace [kube-system], Hybrid mode [true]"
time="2021-07-07T08:29:25Z" level=info msg="Beginning cluster membership, namespace [kube-system], lock name [plndr-svcs-lock], id [k8s01]"
I0707 08:29:25.158970       1 leaderelection.go:243] attempting to acquire leader lease kube-system/plndr-svcs-lock...
time="2021-07-07T08:29:25Z" level=info msg="Beginning cluster membership, namespace [kube-system], lock name [plndr-cp-lock], id [k8s01]"
I0707 08:29:25.159337       1 leaderelection.go:243] attempting to acquire leader lease  kube-system/plndr-cp-lock...
I0707 08:29:28.081058       1 leaderelection.go:253] successfully acquired lease kube-system/plndr-svcs-lock
time="2021-07-07T08:29:28Z" level=info msg="Beginning watching services for type: LoadBalancer in all namespaces"
I0707 08:29:28.081903       1 leaderelection.go:253] successfully acquired lease kube-system/plndr-cp-lock
time="2021-07-07T08:29:28Z" level=info msg="This node is starting with leadership of the cluster"
time="2021-07-07T08:29:28Z" level=info msg="Broadcasting ARP update for 192.168.0.40 (00:50:56:a5:4f:05) via ens160"
time="2021-07-07T08:29:28Z" level=info msg="Node [k8s01] is assuming leadership of the cluster"
time="2021-07-07T08:29:28Z" level=info msg="Service [kubernetes] has been addded/modified, it has no assigned external addresses"
time="2021-07-07T08:29:28Z" level=info msg="Service [kube-dns] has been addded/modified, it has no assigned external addresses"
time="2021-07-07T08:29:31Z" level=info msg="Broadcasting ARP update for 192.168.0.40 (00:50:56:a5:4f:05) via ens160"
time="2021-07-07T08:29:34Z" level=info msg="Broadcasting ARP update for 192.168.0.40 (00:50:56:a5:4f:05) via ens160"
time="2021-07-07T08:29:37Z" level=info msg="Broadcasting ARP update for 192.168.0.40 (00:50:56:a5:4f:05) via ens160"
time="2021-07-07T08:29:40Z" level=info msg="Broadcasting ARP update for 192.168.0.40 (00:50:56:a5:4f:05) via ens160"

...

time="2021-07-07T08:38:03Z" level=info msg="Broadcasting ARP update for 192.168.0.40 (00:50:56:a5:4f:05) via ens160"
time="2021-07-07T08:38:06Z" level=info msg="Broadcasting ARP update for 192.168.0.40 (00:50:56:a5:4f:05) via ens160"
time="2021-07-07T08:38:09Z" level=info msg="Broadcasting ARP update for 192.168.0.40 (00:50:56:a5:4f:05) via ens160"
time="2021-07-07T08:38:12Z" level=info msg="Broadcasting ARP update for 192.168.0.40 (00:50:56:a5:4f:05) via ens160"
time="2021-07-07T08:38:15Z" level=info msg="Broadcasting ARP update for 192.168.0.40 (00:50:56:a5:4f:05) via ens160"
time="2021-07-07T08:38:19Z" level=info msg="Broadcasting ARP update for 192.168.0.40 (00:50:56:a5:4f:05) via ens160"
time="2021-07-07T08:38:21Z" level=info msg="Service [nginx-dhcp] has been addded/modified, it has an assigned external addresses [0.0.0.0]"
time="2021-07-07T08:38:21Z" level=info msg="Creating new macvlan interface for DHCP [vip-e271d179]"
time="2021-07-07T08:38:21Z" level=info msg="Service [nginx-dhcp] has been addded/modified, it has an assigned external addresses [0.0.0.0]"
time="2021-07-07T08:38:22Z" level=info msg="Broadcasting ARP update for 192.168.0.40 (00:50:56:a5:4f:05) via ens160"
time="2021-07-07T08:38:22Z" level=info msg="DHCP VIP [192.168.0.121] for [nginx-dhcp/e271d179-b9cf-44d1-bc9f-b076bead9403] "
time="2021-07-07T08:38:22Z" level=info msg="Starting advertising address [192.168.0.121] with kube-vip"
time="2021-07-07T08:38:22Z" level=info msg="Started Load Balancer and Virtual IP"
time="2021-07-07T08:38:22Z" level=info msg="Updating service [nginx-dhcp], with load balancer address [192.168.0.121]"
time="2021-07-07T08:38:22Z" level=info msg="Broadcasting ARP update for 192.168.0.121 (00:50:56:a5:4f:05) via ens160"
time="2021-07-07T08:38:22Z" level=info msg="Service [nginx-dhcp] has been addded/modified, it has an assigned external addresses [0.0.0.0]"
time="2021-07-07T08:38:25Z" level=info msg="Broadcasting ARP update for 192.168.0.40 (00:50:56:a5:4f:05) via ens160"
time="2021-07-07T08:38:25Z" level=info msg="Broadcasting ARP update for 192.168.0.121 (00:50:56:a5:4f:05) via ens160"
time="2021-07-07T08:38:28Z" level=info msg="Broadcasting ARP update for 192.168.0.40 (00:50:56:a5:4f:05) via ens160"

dan@k8s01:~$ ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: ens160: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 00:50:56:a5:4f:05 brd ff:ff:ff:ff:ff:ff
    inet 192.168.0.41/24 brd 192.168.0.255 scope global ens160
       valid_lft forever preferred_lft forever
    inet 192.168.0.40/32 scope global ens160
       valid_lft forever preferred_lft forever
    inet 192.168.0.121/32 scope global ens160
       valid_lft forever preferred_lft forever
    inet6 fe80::250:56ff:fea5:4f05/64 scope link
       valid_lft forever preferred_lft forever
3: docker0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default
    link/ether 02:42:3b:ad:92:b0 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.1/16 brd 172.17.255.255 scope global docker0
       valid_lft forever preferred_lft forever
6: tunl0@NONE: <NOARP,UP,LOWER_UP> mtu 1480 qdisc noqueue state UNKNOWN group default qlen 1000
    link/ipip 0.0.0.0 brd 0.0.0.0
    inet 10.0.73.64/32 scope global tunl0
       valid_lft forever preferred_lft forever
7: calia2f764bc6fa@if4: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1480 qdisc noqueue state UP group default
    link/ether ee:ee:ee:ee:ee:ee brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet6 fe80::ecee:eeff:feee:eeee/64 scope link
       valid_lft forever preferred_lft forever
8: cali4ce821b3157@if4: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1480 qdisc noqueue state UP group default
    link/ether ee:ee:ee:ee:ee:ee brd ff:ff:ff:ff:ff:ff link-netnsid 1
    inet6 fe80::ecee:eeff:feee:eeee/64 scope link
       valid_lft forever preferred_lft forever
9: vip-e271d179@ens160: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default
    link/ether 96:65:01:45:0e:23 brd ff:ff:ff:ff:ff:ff
    inet6 fe80::9465:1ff:fe45:e23/64 scope link
       valid_lft forever preferred_lft forever
dan@k8s01:~$ curl 192.168.0.121
<!DOCTYPE html>
<html>
<head>
<title>Welcome to nginx!</title>
```

**Comment by thebsdbox** [comment]
tenor

**Pull request merged**